### ISSUE TITLE
[Bug Report] Numerical Instability in ttnn.softmax

### ISSUE THREAD

**Issue by cthsieh** [tenstorrent/tt-metal#13420]
**Describe the bug**

The function `ttnn.softmax` exhibits numerical instability when handling input tensors with large values. Specifically, when provided with such inputs, the function produces incorrect results due to overflow, rendering the output tensor inaccurate.

**To Reproduce**

The following example code demonstrates the bug:

```python
import torch
import ttnn

x = torch.tensor([  213.0000,   159.0000,   108.5000, -1208.0000,    94.0000,
                    169.0000,   171.0000,    40.2500,    34.0000,    33.0000,
                    1216.0000,    33.0000,    21.8750,    36.2500,   141.0000,
                    2384.0000,    70.5000,   221.0000,   175.0000,   103.5000,
                    170.0000,    37.0000,    37.7500,    60.7500,    56.7500,
                    25.1250,    57.7500,    26.7500]).reshape(1, 28, 1, 1)
golden = torch.softmax(x, dim=-1)
print(f'golden = {golden.reshape(28)}')

device=ttnn.open_device(device_id=0)

x_tt = ttnn.as_tensor(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, memory_config=ttnn.L1_MEMORY_CONFIG, device=device)
y_tt = ttnn.softmax(x_tt, dim=-1)
result = ttnn.to_torch(y_tt)
print(f'result = {result.reshape(28)}')

ttnn.close_device(device)
```

This issue is critical because the input tensor values above originate from a real-world model, [Qwen2](https://huggingface.co/Qwen/Qwen2-7B), and the problem affects the correctness of the model's computations.

**Expected behavior**

By running the above example, you should find the expected result is
```
golden = tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```

This softmax computation is performed over the last dimension, which has a length of 1. As a result, each element is the largest in its own scope and is normalized to the maximum possible softmax value of 1.0.

However, the actual computed result is
```
result = TorchTensor(
[0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
 0.0000, 0.0000, 0.9922, 0.9922, 0.9961,
 0.0000, 0.9961, 0.9961, 0.9961, 0.0000,
 0.0000, 0.9961, 0.0000, 0.0000, 0.0000,
 0.0000, 0.9961, 1.0000, 0.9961, 0.9844,
 0.9961, 1.0000, 1.0000], dtype=torch.bfloat16)
```

By comparing the above result with the input tensor position by position, we observe that when the input value is large (greater than 100), the result is incorrectly computed as zero.

**Please complete the following environment information:**
- OS: Ubuntu 20.04
- Version of software: e9078f9975b9c96c27fad49aba8289de17c452e6

**Additional context**

The exponential computation in the Softmax function can result in very large numbers, especially when the inputs are large. To address this, a common trick is to subtract the maximum value in the input vector before applying the exponential function. This makes the computation more stable without changing the result.

**Comment by yugaoTT** [comment]
please try with numeric_stable=1 option when calling softmax

**Comment by ayerofieiev-tt** [comment]
@yugaoTT , this option should be enabled by default. Operation should work to start with. 
Then any optimization should be made on top. Can you please check on this with Jasmina and make a change?

**Comment by dmakoviichuk-tt** [comment]
math_approx enabled cost us a week of pain while we tried to find a reason why our code diverges. 
It is a best example.

**Comment by cthsieh** [comment]
> please try with numeric_stable=true option when calling softmax

Yes, this solves my problem. Thanks.

BTW, I don't understand why we wouldn't enable this numerically stable option by default. It ensures that the softmax function works correctly under all input conditions, which should be a fundamental requirement for any implementation.

> just as we enable math_approx (less precision) by default

How to disable it? I'm really struggling with the precision issue while implementing a model using ttnn. I want to disable it to make debugging more manageable.

**Comment by yugaoTT** [comment]
Yeah that's because we only very recently added the support for numeric stable, so it's set to default false to keep the models Same perf.

**Comment by yugaoTT** [comment]
you can disbale it by set comute_kernel_config, see examples in `tests/tt_eager/python_api_testing/unit_testing/misc/test_layernorm_sharded.py`
